Add EIP: Hardware and Bandwidth Recommendations

kevaundray opened the pull request:
Original documents are here (for validators):

- Hardware: https://hackmd.io/G3MvgV2_RpKxbufsZO8VVg?view
- Bandwidth: https://hackmd.io/DsDcxDAVShSSLLwHWdfynQ?view

kevaundray committed:
draft

kevaundray commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,147 @@
+
+### Bandwidth
+
+- Statista states that as of January 2024:
+  - The global average download for broadband is 92 Mbps and the global average upload is 43 Mpbs.
+  - The global average download for mobile is 50 Mbps and 11 Mbps
+- GSMA report showing the state of mobile internet connectivity in 2024 shows that:
+  - The mobile upload speeds in Higher Income Countries (HIC) is about 18 Mbps
+  - The global average mobile download is 48 Mbps.
+
```
GSMA link is https://www.gsma.com/r/wp-content/uploads/2024/10/The-State-of-Mobile-Internet-Connectivity-Report-2024.pdf?utm_source=website&utm_medium=button&utm_campaign=somic24

Paghe 39 for the upload speeds of HIC and page 5 for the 48 Mbps figure.

Statista link is: https://www.statista.com/statistics/896779/average-mobile-fixed-broadband-download-upload-speeds/

kevaundray replied:
Let me know how you would want these to be referred to in the document

kevaundray commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,147 @@
+
+## Backwards Compatibility
+
+This EIP is informational and requires no protocol changes. We recommend that future EIPs include an assessment of their impact on these hardware requirements.
```
Would the recommendation here require another EIP if this gets finalized?

In any case, I don't think it belongs here since this section is explicitly about whether the EIP breaks backwards compatibility

kevaundray committed:
Update EIPS/draft_hardware_and_bandwidth_recs.md

kevaundray committed:
add rationale for 64GB RAM (only for validators)

kevaundray commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,148 @@
+
+RAM/memory is dominated by state cache. As of January 2025, it is possible to run a full node with 16GB of RAM, however this has been known to not work with all combinations of EL and CL clients in the past.
+
+On 32GB vs 64GB; 32GB works right now, however we recommend 64GB as [preliminary benchmarks](https://hackmd.io/@han/bench-hash-in-snark) have shown that zk-STARKS can consume a significant amount of memory and the difference in cost relative to the entire hardware setup for a validator is insignificant.
```
I saw that EIPs 2926, 2938, 3298, 3416 and 3607 also used hackmd links however the contents of them can easily be changed, so I wonder if there is a rule about using them?

kevaundray committed:
Copy over authors

kevaundray committed:
use simpler word

potuz commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,148 @@
+
+## Abstract
+
+This proposal specifies hardware and bandwidth requirements for full nodes and validators. Validators are split into attesters and proposers. The resource heavy component of proposing a block is creating it and broadcasting the necessary data needed to ensure that other attesters view the block as valid. We use the "block builder" to further emphasize this.
```
The title and description of the EIP use the word "recommendation". The title description in the PR and the very first sentence of the abstract use the word "requirement". 

Throughout the document these two are used interchangeably and it's not clear what's a recommendation or what is a requirement. For benchmarking reasons or for informed decision-making, is Ethereum meant to use the required system or the recommended one?

kevaundray replied:
Will clarify/clean up terminology in the doc tomorrow -- Essentially you can run your node with less than what is listed in the document and it may work fine, but benchmarks etc would be based off of what is listed in there. So merging this PR does not imply immediate changes are needed to ones setup

smartprogrammer93 reviewed (commented):
Given EIPs are immutable once they become final, does it make sense to place these requirements in an EIP? I expect requirements will need to be updated in a couple of years. In that case, we will need to create a different EIP amending this one. Then, operators will need to check 2 different EIPs to know what the current requirements are. Seems impractical to me.

chfast commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,148 @@
+
+### Recommended hardware and bandwidth specifications
+
+| Node type | Storage | RAM | CPU Cores | CPU Single Thread/Multithread rating | Download/Upload speed |
```
You should explain what is this "rating".

kevaundray replied:
Yep will do

restakeservice commented:
good idea!

kevaundray committed:
add note about passmark

kevaundray commented:
> Given EIPs are immutable once they become final, does it make sense to place these requirements in an EIP? I expect requirements will need to be updated in a couple of years. In that case, we will need to create a different EIP amending this one. Then, operators will need to check 2 different EIPs to know what the current requirements are. Seems impractical to me.

In the call someone mentioned that, the status of this could be changed to "live" and we modify it in-place instead of creating a new EIP each time. I would defer to the EIP maintainers regarding that and whats possible there.

I agree that creating a new EIP whenever we change specs is undesirable.

kevaundray committed:
requirements -> recommendations

suchapalaver commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,152 @@
+
+## Abstract
+
+This proposal specifies hardware and bandwidth recommendations for full nodes and validators. Validators are split into attesters and proposers. The resource heavy component of proposing a block is creating it and broadcasting the necessary data needed to ensure that other attesters view the block as valid. We use the "block builder" to further emphasize this.
```
```suggestion
This proposal specifies hardware and bandwidth recommendations for full nodes and validators. Validators are split into attesters and proposers. The resource-intensive aspect of block proposal lies in creating the block and broadcasting the data required for attesters to validate it. We use the "block builder" to further emphasize this.
```

kevaundray replied:
Github is not allowing to apply the changes because another change was added -- I will add these in the cli and put you as co-author

suchapalaver commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,152 @@
+
+This proposal specifies hardware and bandwidth recommendations for full nodes and validators. Validators are split into attesters and proposers. The resource heavy component of proposing a block is creating it and broadcasting the necessary data needed to ensure that other attesters view the block as valid. We use the "block builder" to further emphasize this.
+
+We note that you may be able to run a client with less than the recommended specifications, however benchmarks and decision-making will be made with respects to these recommendations.
```
```suggestion
We note that you may be able to run a client with less than the recommended specifications, however benchmarks and decision-making will be made with respect to these recommendations.
```

suchapalaver commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,152 @@
+
+- Ensuring meaningful benchmark comparisons across different implementations
+- Enabling informed decision-making about protocol upgrades and their hardware implications
+- Providing clear guidance for node operators with respects to the future
```
```suggestion
- Providing clear guidance for node operators with respect to the future
```

suchapalaver commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,152 @@
+Without a shared understanding of target hardware specifications:
+
+- Benchmark results lose significance due to inconsistent testing environments
+- Decision-making becomes challenging for implementation choices, as performance characteristics are heavily hardware dependent
```
```suggestion
- Decision-making becomes challenging for implementation choices, as performance characteristics are heavily hardware-dependent
```

suchapalaver commented on EIPS/draft_hardware_and_bandwidth_recs.md:
```diff
@@ -0,0 +1,152 @@
+| Attester      | 4TB     | 64GB    | 8 cores/16 threads | 3500 / 25000 | 50 Mbps / 25 Mbps |
+| Block builder | 4TB     | 64GB    | 8 cores/16 threads | 3500 / 25000 |  100 Mbps / 50 Mbps |
+
+- The CPU rating is a measure of how powerful a particular CPU is for single and multithreaded respectively. All numbers references are normalized according to PassMark.
```
```suggestion
- The CPU rating is a measure of how powerful a particular CPU is for single and multithreaded respectively. All numbers referenced are normalized according to PassMark.
```

OisinKyne commented:
There doesn't seem to be a `discussions-to`, so commenting here for now:

What might help to minimise future debate/bikeshedding over what the new minimum should be is a price based guideline to help keep track of the performance growth in what 'commodity hardware' might mean. E.g. what $1000 capex + $100 p/a opex gets you, or what x% of a validator price can get you, or what y% of a validator's annual reward can get you.

SamWilsn committed:
Update and rename draft_hardware_and_bandwidth_recs.md to eip-7870.md

timbeiko commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,153 @@
+Clear system specifications are crucial for:
+
+- Ensuring meaningful benchmark comparisons across different implementations
+- Enabling informed decision-making about protocol upgrades and their hardware implications
```
Both HW and bandwidth

```suggestion
- Enabling informed decision-making about protocol upgrades and their resource usage implications
```

timbeiko commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,153 @@
+
+## Abstract
+
+This proposal specifies hardware and bandwidth recommendations for full nodes and validators. Validators are split into attesters and proposers. The resource heavy component of proposing a block is creating it and broadcasting the necessary data needed to ensure that other attesters view the block as valid. We use the "block builder" to further emphasize this.
```
Consider renaming "block builder" to "solo" or "local" block builder to emphasize the distinction with professional block builders.

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Tim Beiko <9390255+timbeiko@users.noreply.github.com>

kevaundray committed:
Co-authored-by: Joseph Livesey <joseph@semiotic.ai>

kevaundray committed:
block builder -> local block builder

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Mercy Boma Naps Nkari <96525594+bomanaps@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Mercy Boma Naps Nkari <96525594+bomanaps@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Mercy Boma Naps Nkari <96525594+bomanaps@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Mercy Boma Naps Nkari <96525594+bomanaps@users.noreply.github.com>

kevaundray committed:
remove requires field

dapplion commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+- Samsung 990 Pro
+- Seagate Firecuda 530
+- Teamgroup MP44
+- WD Black SN850X
```
Do you have to list specific commercial models in an EIP? Can you focus of i/o metrics that are brand agnostic?

advaita-saha replied:
indeed i/o metrics would be more generalised and open up more broader choices for people

kevaundray replied:
I took the weakest SSD brand in terms of random reads and writes and used its i/o metrics.

This was the Kingston KC3000 -- all of them were roughly the same.

Link: https://www.kingston.com/unitedkingdom/en/ssd/kc3000-nvme-m2-solid-state-drive?capacity=2048gb

dapplion commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+- Samsung 990 Pro
+- Seagate Firecuda 530
+- Teamgroup MP44
+- WD Black SN850X
```
Do you have to list specific commercial models in an EIP? Can you focus of i/o metrics that are brand agnostic?

kevaundray replied:
Fixed -- see other comment

CPerezz commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+- The CPU rating is a measure of how powerful a particular CPU is for single and multithreaded respectively. All numbers referenced are normalized according to PassMark.
+
+### Recommended Prebuilds (Attester and Local block builder)
```
Even if you suggest different options, could you stick to a single one for the benchmarks in the future? 2 as max if we include RPI which otherwise is definitely non-represented.

Meaning this wouldn't just be the hardware recommendations but also the common denominator for all benchmarks given from now on. 
In that way, is significantly easier to compare cryptographic protocols rather than giving benchmarks on an M4 Max and then having no clue if is suitable for Attesters or block builders etc..

If not fixed to a machine, at least fixed to a number of cores a threads would definitely help.

chfast replied:
It is not good idea nor practical to do benchmarks on single hardware. You can instead tune your CPU to the provided rating.

CPerezz replied:
Mostly say so because depending on the CPU, it might have AVX-{} support or not. Same with custom instruction sets for asm. RPI won't support Intel's fast modular field custom instruction set for example.

These are all issues that adjusting to rating won't support. Also could help to dismiss or incentivize the implementation of such specific backends to squeeze as much perf as possible.

Finally, maybe not as big of an issue (I ignore that part) but different pipelining, cache lines and other kinds of similar properties can also influence some of the stuff we bench. Specifically when we work sometimes within the realm of `us/ns`.

This was the main reason behind the suggestion. But if there are well-known alternatives that solve that or this can also be "adjusted", please let me know to resolve the comment! cc: @chfast

kevaundray replied:
Added note about DRAM cache and TLC flash storage

cshein45 commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+- The CPU rating is a measure of how powerful a particular CPU is for single and multithreaded respectively. All numbers referenced are normalized according to PassMark.
+
+### Recommended Prebuilds (Attester and Local block builder)
```
MOVE

kevaundray commented:
Feedback from EIP maintainers:

- Remove brands (including recommended prebuild), so as to not to favour particular brands
- Add rationale on how the particular numbers were chosen*
- Add hackmd doc to assets folder, so it can be linked in the EIP
- Will need a call to action to vote on the statista and GSMA links
- Change EIP from Draft to "Ready to review"

* These were in the original document, but the supporting evidence for them required a lot of external links.

@SamWilsn just want to confirm, I did not miss anything?

jochem-brouwer commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+  - Single Thread: 3903
+  - Multithread: 30367
+
+#### ASUS NUC 14 Pro
```
An idea, but maybe move those recommended prebuilds to an external site (as part of removing the brand names). HackMD will do fine. If there is some kind of external site which actually document and review setups that would also be great (but not aware of any).

kevaundray replied:
This makes sense -- I think people will want some document that has more precise suggestions. It could be a hackmd that says something like "concrete suggestions for EIP XXXX"

jochem-brouwer commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+For attesters, we recommend: 50 Mbps download and 25 Mbps upload.
+
+The minimum value recorded to work as of January 2025 has been ~15 Mbps. This was recorded in simulations done by Eth panda ops where 40% of the network had Gigabit connections and the other 60% had 15 Mbps upload.
```
```suggestion
The minimum value recorded to work as of January 2025 has been ~15 Mbps. This was recorded in simulations done by ethPandaOps where 40% of the network had Gigabit connections and the other 60% had 15 Mbps upload.
```

Not sure if this is the desired casing, but took this from their site.

jochem-brouwer commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+## Specification
+
+### Recommended hardware and bandwidth specifications
```
Maybe include a date here, so if users stumble upon this EIP they will know that this is recent, and not recommended specifications of 3 years ago.

tynes replied:
We should keep this EIP up to date or explicitly replace it with another in the future

jochem-brouwer commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+This EIP is informational and requires no protocol changes. We recommend that future EIPs include an assessment of their impact on these hardware recommendations.
+
+## Security Considerations
```
I think I'd like to see a section about "Future". What could we expect in the future? Would we need more storage? If I build a node using the recommended settings, what is the target duration that this setup is usable and needs no hardware upgrades?

kevaundray replied:
Off the top of my head, this is what I can think of. Kind of hard to be comprehensive and I'm sure there are nuances that are missed or mistakes:

**Full node**

- Download and upload speed goes down once we have peerDAS
- Storage will go down once we have EIP444
- Storage and RAM will go down once we have stateless (unclear how stateless clients will do state cache for re-org protection)
- CPU could go down further once we snarkify the EL block
- CPU and storage might go up depending on how far we increase the gas limit and whether we have EIP4444 and stateless

**Attesters**

- They inherit all of the changes from full nodes (though attesters sample more, so higher bandwidth)
- Download speed goes up due to increase in blobs

**Block builder**

- CPU usage goes up once we have stateless due to the need to create the stateless proofs
- Download and upload speed initially goes down to peerDAS
- Upload speed and cpu requirements goes down due to epbs due to having more time to create the execution payload
- Download and upload speed goes up due to increase in blobs
- CPU requirements increase due to the addition of heavier proofs like zkvm (probably will need a GPU)

 (One note here is that until we have zkvm proofs, you can still block build but at a lower throughput by setting the maximum number of blobs you can do in your EL based on your bandwidth requirements) 

I think the main entity that will change significantly is the block builder.

Other changes to consider:

- State network: This refers to the portal state network -- it would in theory, allow us to increase the gas limit and so it would increase builder requirements and the download speed of attesters and full nodes. 

The argument here is that if we increase the gas limit, we increase the rate of state growth and eventually no-one but centralized parties will be able to hold the state and they could deny access or censor the rest of the network. But with state network, we would always be able to retrieve the state (at a slower throughput).
  
- Delayed state root: depending on your client and how long it was taking to compute the state root, the builder cpu requirements can go down

- FOCIL; This effectively puts a lower bound on the block size and I think mostly affects block builders, in practice, I think the FOCIL limit will be negligible and won't affect the specs.

- Rainbow staking derivatives: This will add more roles -- separating the centralizing components like block building from the other components at the very least. 

It has been argued that this might give folks reason to pump up the requirements for the centralizing components even further like block building. I am unsure about how this will play out.

jochem-brouwer reviewed (commented):
Some questions! :smile: :+1:

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Jochem Brouwer <jochembrouwer96@gmail.com>

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+  - Single Thread: 3520
+  - Multithread: 25158
+
+### Recommended Prebuild (Full node)
+
+#### Raspberry Pi 5
+
+- 4 core
+- 4 threads
+- CPU Rating
+  - Single thread: 1487
+  - Multithred: 3428
```
I'm a little concerned that a RP5 would be unable to function as a full node. Can anyone confirm this works? For one, storage speeds are quite important and it only operates at PCIe 2.0 speeds (like ~500MB/s). This might be do-able, but it's not something I would recommend today. Instead, I would suggest a Rock 5 Model B which is all-around better.

jtraglia replied:
Apparently a RP5 works. Very cool. Consider my concern withdrawn.

See: https://x.com/EthereumOnARM/status/1880258026676146618

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+- 4 threads
+- CPU Rating
+  - Single thread: 1487
+  - Multithred: 3428
```
```suggestion
  - Multithread: 3428
```

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+- Teamgroup MP44
+- WD Black SN850X
+
+In particular, we recommend purchasing NVMe M.2 instead of SATA as NVMe has higher throughput.
```
Additionally, we should mention that we recommend NVMe drives with DRAM and TLC flash. QLC flash is slower and has a lower write endurance, meaning they may fail sooner.

kevaundray replied:
Added note about DRAM cache and TLC flash storage

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+For validators that want to build blocks locally, we recommend global bandwidth figures inline with the global average for fixed broadband: 100 Mbps download and 50 Mbps upload.
+
+*If a block builder does not have the recommended bandwidth, we recommend that they build a block that is partially full and or one that includes less blobs.*
```
```suggestion
*If a block builder does not have the recommended bandwidth, we recommend that they build a block that is partially full and or one that includes fewer blobs.*
```

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+As of January 2025, the storage requirements for all node types are dominated by the need to store block history.
+A 2TB disk can be used; however, this is not recommended due to the rate of history growth and the uncertainty as to when EIP-4444 will be completed.
+
+### RAM
+
+RAM/memory is dominated by state cache. As of January 2025, it is possible to run a full node with 16GB of RAM, however this has been known to not work with all combinations of EL and CL clients in the past.
```
Just a nit. I would prefer if the section header were "Memory" and replace "RAM" with "memory" in the paragraph. At the beginning, we could do "Memory (RAM) is dominated by..."

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+### Recommended hardware and bandwidth specifications
+
+| Node type | Storage | RAM | CPU Cores | CPU Single Thread/Multithread rating | Download/Upload speed |
+| -------- | -------- | -------- | --------| -------- |--------|
+| Full node     | 4TB     | 32GB    | 4 cores/8 threads  | 1000 / 3000 | 50 Mbps / 15 Mbps |
+| Attester      | 4TB     | 64GB    | 8 cores/16 threads | 3500 / 25000 | 50 Mbps / 25 Mbps |
+| Local block builder | 4TB     | 64GB    | 8 cores/16 threads | 3500 / 25000 |  100 Mbps / 50 Mbps |
```
Can we format this table so everything is aligned? So it's easier to read when viewing markdown source.

jtraglia replied:
Also, for consistency with the other fields, we should add spacing around "/" in the CPU cores column. And "cores" should be lowercase in the column title.

kevaundray replied:
Yep, for future reference, I used: https://codebeautify.org/markdown-formatter#

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+  - Single Thread: 3520
+  - Multithread: 25158
+
+### Recommended Prebuild (Full node)
```
To match the order of the recommended hardware table, the full node section should be first.

kevaundray replied:
Since we don't want brands, I'll remove this whole section on recommended prebuilds

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+- Statista states that as of January 2024:
+  - The global average download for broadband is 92 Mbps and the global average upload is 43 Mbps.
+  - The global average download for mobile is 50 Mbps and 11 Mbps
```
Missing punctuation.
```suggestion
  - The global average download for mobile is 50 Mbps and 11 Mbps.
```

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+  - The global average download for broadband is 92 Mbps and the global average upload is 43 Mbps.
+  - The global average download for mobile is 50 Mbps and 11 Mbps
+- GSMA report showing the state of mobile internet connectivity in 2024 shows that:
+  - The mobile upload speeds in Higher Income Countries (HIC) is about 18 Mbps
```
Missing punctuation.
```suggestion
  - The mobile upload speeds in Higher Income Countries (HIC) is about 18 Mbps.
```

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,152 @@
+
+#### Attesters
+
+For attesters, we recommend: 50 Mbps download and 25 Mbps upload.
```
The colon isn't really necessary here.
```suggestion
For attesters, we recommend 50 Mbps download and 25 Mbps upload.
```

kevaundray committed:
format table

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

cshein45 reviewed (approved):
MOVEMENT

kevaundray committed:
RAM -> memory

kevaundray committed:
replace branded SSD with i/o metrics

kevaundray commented:
@yorickdowne @potuz Thanks for the comments on hardware -- any comments on bandwidth?

kevaundray committed:
add note about TLC and DRAM

kevaundray committed:
remove recommended prebuilds

0xFredrik committed:
Update eip-7870.md

Made adjustments to make it a bit easier to read.

Included information about how even if you use mev-boost you may need the same bandwidth as a local block builder under certain conditions.

Included some additional details for choosing a NVMe drive (such as avoiding QLC and DRAMless drives).

Added some information about what the PassMark CPU ratings translate into

Added a CPU section

Removed Statista as I don't believe it serves much of a purpose in the EIP, and I'm not sure how accurate these numbers really are when looking at the median.

Added Quick Reference for those who just want the details

kevaundray committed:
Merge pull request #1 from fredriksvantes/kw/hardware-bandwidth-recs

Update eip-7870.md

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,146 @@
+|-------------------------|---------------|-------|---------------------|----------------------------------------------------|--------------------------|
+| **Full Node**           | 4 TB NVMe     | 32 GB | 4c / 8t             | ~1000 ST / 3000 MT                                 | 50 Mbps / 15 Mbps        |
+| **Attester**            | 4 TB NVMe     | 64 GB | 8c / 16t            | ~3500 ST / 25000 MT                                | 50 Mbps / 25 Mbps        |
+| **Local Block Builder** | 4 TB NVMe     | 64 GB | 8c / 16t            | ~3500 ST / 25000 MT                                | 100 Mbps / 50 Mbps       |
```
The table needs to be re-formatted again 😄

jtraglia replied:
Also, I think we should change "RAM" in the title to "Memory".

jtraglia replied:
Also "speed" in "Download / Upload speed" is lowercase, when it probably shouldn't be. I'd be in favor of renaming this title to "Bandwidth Download / Upload".

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,146 @@
+### CPU
+
+- **Single vs. Multi-thread Performance**
+  - Attesters and local block builders benefit from both high single-thread performance and solid multi-thread performance.  
```
The word "solid" stands out here. Feels too informal. I'd probably rephrase somehow. Maybe:
```suggestion
  - Attesters and local block builders benefit from both high single-thread and multi-thread performance.  
```
Also, this statement doesn't really compare ("single vs multi") or tell why it's important. Just that it is.

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,146 @@
+- **Rationale**:
+  - Most MEV-relay interactions involve fetching bundles and block headers, which can be done within typical broadband speeds.
+  - While the local validator will also share the block with its peers, the relay will do the same which reduces the need for local bandwidth.
+  - There may however be cases where your validator will still build a local block, such as if no mev-relay responds or if the value of the mev reward is lower than the minimum configuration set in mev-boost. In these circumstances, the recommendations for **Local Block Builders** would be relevant.
```
```suggestion
  - However, there may be cases where your validator will still build a local block, such as if no MEV-relay responds or if the value of the MEV reward is lower than the minimum configuration set in MEV-Boost. In these circumstances, the recommendations for **Local Block Builders** would be relevant.
```
* We're missing commas around "however" but it flows better if we move it to the front.
* We use `MEV-relay` in another point. For consistency, we should capitalize MEV in other places too.
  * The "B" in boost should be capitalized too. See: https://github.com/flashbots/mev-boost#what-is-mev-boost

jtraglia replied:
Btw, there are other instances of "mev" -> "MEV" which need to be updated.

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,146 @@
+- **Recommended**: 50 Mbps download / 15 Mbps upload.
+- **Rationale**:
+  - Full nodes currently participate in sampling and must track the chain tip, but are not as latency-sensitive as attesters or local block builders.
+  - They can operate on lower bandwidth but risk being a slot or more behind the chain if capacity is severely limited.
```
Just to be extra clear. I thought this was about system resources while skimming 🤦
```suggestion
  - They can operate on lower bandwidth but risk being a slot or more behind the chain if bandwidth capacity is severely limited.
```

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,146 @@
+  - **Storage**: 4 TB NVMe  
+  - **RAM**: 64 GB  
+  - **CPU**: 8 cores / 16 threads (~3500 ST / ~25000 MT PassMark)  
+  - **Bandwidth**: 100 Mbps down / 50 Mbps up  
+
+
+## Backwards Compatibility
```
An extra blank line can be deleted.
```suggestion
  - **Bandwidth**: 100 Mbps down / 50 Mbps up  

## Backwards Compatibility
```

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
multi:

- format table
- "ram" - "memory"
- "Download upload speed" -> "Bandwidth Download / Upload"

kevaundray committed:
mev-boost -> MEV-Boost

kevaundray committed:
grammar

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kamilchodola commented:
General question to this one:

Should we better treat that as "Recommended setup" or "Minimal setup"?

Minimal would be then something we can refer to and adjust to it - having any users on our support lines which will be struggling with weaker setups could be then advised to think of any improvement and referred to this EIP.
Recommendation always gives quite a bit of room for complains. 

Also having "Minimal supported setup" in place we can adjust any benchmarking tools to it and then make a proper assumptions.

kamilchodola commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+
+Node operators typically run both an **Execution Layer (EL)** client and a **Consensus Layer (CL)** client on the same machine. The specifications below assume the combined resource usage of both.
+
+| Node Type               | Storage   | Memory | CPU Cores / Threads | **PassMark CPU Rating** | Bandwidth Download / Upload |
```
Are all those numbers applied based on some research made? Or based on experience from discussions with stakers?

From Nethermind perspective those are super comfortable.

Also shall we distinguish EL and CL requirements to be treated separately? I saw a lot of setups where actually both are running on different machines.

kevaundray replied:
I think the average user would likely run these on the same machine? 

The rationale for them is in the linked hackmd in the comment box, we could not include all of those links so the EIP has been somewhat watered down

kamilchodola commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+
+This EIP is informational and requires no protocol changes. We recommend that future EIPs include an assessment of their impact on these hardware recommendations.
+
+## Security Considerations
```
One thing I'd suggest doing prior to applying that is to do any research among community either by pool on X or anything else which can have bigger outreach.

If it will appear that 51% of users have weaker than specified setup (which is super unlikely but still) we could potentially open a route for some unexpected attacks because we will test on too good hardware.

At least worth to know in case of noticed issue which affect weaker setups how much network (in terms of nodes) is put in danger.

kevaundray replied:
We could alternatively, test against the old NUCs for a year too though the current idea was that instead of us polling N validators, we would give them a tool to allow them to assess themselves and decide when they would want to update

taxmeifyoucan commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+
+| Node Type               | Storage   | Memory | CPU Cores / Threads | **PassMark CPU Rating** | Bandwidth Download / Upload |
+| ----------------------- | --------- | ------ | ------------------- | ----------------------- | --------------------------- |
+| **Full Node**           | 4 TB NVMe | 32 GB  | 4c / 8t             | ~1000 ST / 3000 MT      | 50 Mbps / 15 Mbps           |
```
Considering history expiry, 4TB sounds like a lot to recommend just for a full node. I see it's elaborated on below but I would either provide an estimation how long it will sustain with 2 or 4tb.

taxmeifyoucan replied:
I would also make it clear that these are recommended, not minimum viable specs

taxmeifyoucan replied:
Same goes for 32GB, I believe 16 is enough for different client pairs I have been running. Maybe 24 could be more realistic than 32

kamilchodola replied:
IMO nice to work on 16 or even 8gb but this EIP is some commitment of supported hardware and some guidelines for users on what to use - I'd prefer users to use better hardware especially cost difference (especially for RAM) is like 50-100 bucks to upgrade from 16 to 32 and it gives a lot and makes us a bit more secure from some heavy blocks especially if we consider an increase of their size.

jtraglia replied:
> Considering history expiry, 4TB sounds like a lot to recommend just for a full node. I see it's elaborated on below but I would either provide an estimation how long it will sustain with 2 or 4tb.

Yeah, we recognize that 2TB will be do-able with history expiry, but it's uncertain when that will actually ship. I would prefer the EIP's recommendations be on the safe side, 4TB.

> I would also make it clear that these are recommended, not minimum viable specs

The word "recommendation" is used several times in this doc. It's even in the title:

https://github.com/ethereum/EIPs/blob/69d6453753bc9c2ac36ee1fc0efda193561fef6d/EIPS/eip-7870.md?plain=1#L3

> Same goes for 32GB, I believe 16 is enough for different client pairs I have been running. Maybe 24 could be more realistic than 32

Yes, this 16GB is enough for today (and what my validator has) but might not be enough 3 years from now. Once again, this EIP does not state that you cannot run a full-node with less, just that 32GB is recommended.

taxmeifyoucan replied:
> Yeah, we recognize that 2TB will be do-able with history expiry, but it's uncertain when that will actually ship. I would prefer the EIP's recommendations be on the safe side, 4TB. 

Afaik it's planned for first half of this year so pretty soon. But even if it ships later on, it should still be way before we hit 2TB limit. It's good to be safe, I agree, but let's be clear it's for long term and provide an estimation of how long it will actually take to fill the drive. With current spike in hardware prices, it might be more economic to wait and upgrade later then just buying 4tb immediately. 

I understand the doc is talking about recommended specs but doesn't articulate directly that there is difference between minimum viable and recommend for long term. In this case, the word recommended can be pretty relative and subjective. For someone who only wants to run a full node at home on a low budget, I would not recommend these values. And for an infrastructure provider I would suggest maybe even something even beefier. 
I think the reasoning below is a great insight but this table or motivation above should make it clear what is the target for these recommendations. Or maybe even split the table into Minimum/Recommended as software projects do when listing requirements

kamilchodola replied:
Setting minimum supported hardware for me is super beneficial and already mentioned that few comments above.

Will make it clear under which setups we can't put our sign and which ones we don't believe are good enough.

jtraglia replied:
I agree, minimum hardware requirements would be useful too, but it's a much more difficult problem. I wouldn't expect the minimum for all clients to be the same. In my opinion, minimum requirements are be best left to client teams. For example, the big advantage of Nimbus is that it works very well with low-resource hardware. I believe the current recommendations will work for all clients for at least the next three years.

kamilchodola replied:
Hardware will be much better in next 3 years and currently recommended will probably be much cheaper, so it will be even cheaper than some super cheap ones on which any client barely works now.

taxmeifyoucan replied:
I understand it's much harder to determine what are minimum specs and also harder to sustain so I don't believe it needs to part of the table. However settling now on this much room in recommended also makes me think devs might get too comfortable with the extra performance and it might become the new minimum. I believe we are building also for people with less hardware options. 

@kamilchodola I wouldn't be so sure about significant change in the hardware getting cheaper. That assumes people need to upgrade to latest hardware, their existing nodes are not going to get stronger, only rustier. And I am not very bullish on hardware getting cheaper in general, last few years are proving otherwise and although inflation is slowing down, the progress in hardware is much smaller than it used to be, Moore's law is dead and we cannot count on getting twice as performance in couple of years. Current hardware generations offer rather incremental improvements for high prices

kevaundray replied:
I believe we planned to prune pre-merge data in the first half of the year, which is roughly half a terrabyte.

I got this number by computing the total size of all of the files here: https://era1.ethportal.net/

Total size: 458,497,023 kB
          = 447,751.00 MB
          = 437.26 GB
 
If we use the history growth numbers from paradigm as an approximation, this buys us roughly a year: https://www.paradigm.xyz/2024/05/how-to-raise-the-gas-limit-2 
 
For full history expiry, the storage occupied will be constant, though I think its still uncertain as to when that officially lands in most clients. If it lands in a few months, I think we can change this number for full nodes and validators -- most validators AFAIK have already updated to 4TB, ethstakers for example recommend this for new validators: https://ethstaker.org/staking-hardware

jtraglia commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+---
+eip: 7870
+title: Hardware and Bandwidth Recommendations for Validators and Full Nodes
```
For the EIP title, I think we should remove "for validators and full nodes" for the following reasons:

1. The current name is too long. The max is 44 characters, this is 68. With that part it is 38 characters.
2. It's sort of implied that we're talking about validators & full nodes. But I suppose we could add recommendations for light clients too. A somewhat ambiguous title would allow us to do that.

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

kevaundray committed:
fix

kevaundray committed:
remove link

kevaundray committed:
add note on tuning cache size

kevaundray commented:
> hey @kevaundray , if you could get the linter/bots to pass, we can merge this PR and then you can continually update this basis discussion and feedback

Should be mergeable now :)

kevaundray commented:
@g11tech bump on merging this

Marchhill commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+  - Running a node involves frequent writes (e.g., database updates, logs). Ensure that the SSD’s Total Bytes Written (TBW) rating is sufficient for multi-year operation.
+- **Capacity Considerations**  
+  - As of January 2025, 2 TB can still work, but daily chain growth makes 4 TB more future-proof.  
+  - While EIP-4444 aims to reduce historical storage requirements, the timeline for EIP-4444 remains uncertain.
```
```suggestion
  - While [EIP-4444](./eip-4444.md) aims to reduce historical storage requirements, the timeline for EIP-4444 remains uncertain.
```

Marchhill commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+
+### Memory
+
+- **Why 64 GB for Validators (Attester / Local Block Builder)?**
```
```suggestion
- **Why 64 GB for Validators (Attesters / Local Block Builders)?**
```

Marchhill commented on EIPS/eip-7870.md:
```diff
@@ -0,0 +1,145 @@
+
+### CPU
+
+- **Single vs. Multi-thread Performance**
```
Maybe this section could be expanded slightly, it doesn't give a clear rationale to me

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Marc <Marchhill@users.noreply.github.com>

kevaundray committed:
Update EIPS/eip-7870.md

Co-authored-by: Marc <Marchhill@users.noreply.github.com>

kevaundray committed:
add people who gathered/analyzed data

g11tech reviewed (approved):
lgtm for draft, whether to anchor in time via title or having it as a living EIP, that call can be taken at moving to review time

0xsarawut-eth reviewed (approved):
https://github.com/MeltedMindz/erc-service-object/blob/main/docs%2FERC-draft.md
